Luau Recap April 2021 (#28)
Roblox/luau: docs/_posts/2021-04-30-luau-recap-april-2021.md

---
layout: single
title:  "Luau Recap: April 2021"
---

Luau is our new language that you can read more about at [https://roblox.github.io/luau](https://roblox.github.io/luau). Another busy month in Luau with many performance improvements.

[Cross-posted to the [Roblox Developer Forum](https://devforum.roblox.com/t/luau-recap-april-2021/).]

> ## Features

Reviewer: ```suggestion
## Editor features
```

> Luau implementation now provides an API for type-aware autocomplete suggestions.

Reviewer: ```suggestion
Luau implementation now provides an internal API for type-aware autocomplete suggestions.
```

* By optimizing the way string buffers are handled internally, we bring improvements to string operations including `string.lower`, `string.upper`, `string.reverse`, `string.rep`, `table.concat` and string concatenation operator `..`. Biggest improvements can be seen on large strings
* Improved performance of `table.insert` and `table.remove`. Operations in the middle of large arrays can be multiple times faster with this change
* Improved performance of internal table resize which brings additional 30% speedup for `table.insert`
* Improved performance of checks for missing table fields

## Generic functions

We had to temporarily disable generic function definitions last month after finding critical issues in the implementation.

While they are still not available, we are making steady progress on fixing those issues and making additional typechecking improvements to bring them back in.

## Debugger improvements

Debugging is now supported for parallel Luau Actors in Roblox Studio.

Read more about the feature [on Roblox Developer forums](https://devforum.roblox.com/t/parallel-lua-beta/) and try it out yourself.

## Behavior changes

> Backwards compatibility is very important for Luau, but sometimes a change is required to improve the language or performance. Even still, we try to keep impact of these changes to a minimum:

Reviewer: ```suggestion
Backwards compatibility is important for Luau, but sometimes a change is required to fix corner cases in the language / libraries or to improve performance. Even still, we try to keep impact of these changes to a minimum:
```
